lhotse/dataset/collation.py

```diff
@@ -1,11 +1,109 @@
-from typing import Iterable, Union
+from typing import Iterable, Union, Literal, Tuple, Callable
+from dataclasses import dataclass
 
 import numpy as np
 import torch
 from torch.nn import CrossEntropyLoss
 
 from lhotse import CutSet
 from lhotse.cut import MixedCut
+from lhotse.utils import LOG_EPSILON
+
+
+@dataclass
+class CollateTokens:
+
+    tokens: Iterable[str]
+    add_eos: bool = True
+    add_bos: bool = True
+    pad_symbol: str = "<pad>"
+    bos_symbol: str = "<bos>"
+    eos_symbol: str = "<eos>"
+    unk_symbol: str = "<unk>"
+
+    def __post_init__(self):
+        tokens_unique = (
+            [self.pad_symbol, self.unk_symbol]
+            + ([self.bos_symbol] if self.add_bos else [])
+            + ([self.eos_symbol] if self.add_eos else [])
+            + sorted(set(self.tokens))
+        )
+        self.token2int = {token: idx for idx, token in enumerate(tokens_unique)}
+
+    def __call__(self, cuts: CutSet) -> Tuple[torch.Tensor, torch.Tensor]:
+        token_sequences = [cut.supervisions[0].text for cut in cuts]
+        max_len = len(max(token_sequences, key=len))
+        seqs = [
+            ([self.bos_symbol] if self.add_bos else [])
+            + seq
+            + ([self.eos_symbol] if self.add_eos else [])
+            + [self.pad_symbol] * (max_len - len(seq))
+            for seq in token_sequences
+        ]
+
+        tokens_batch = torch.Tensor([[self.token2int(token) for token in seq] for seq in seqs], dtype=int)
```

Comment:
this doesn't work for me in torch 1.6.0 (this method of creating a tensor as torch.Tensor([list], dtype=int).
also int appears to map to torch.int64, better to use torch dtype.

Reply:
agreed. Btw I haven't tested all code aspects. This is more like a design idea - a draft PR.

Reply:
fixed